// apeman-react-footer/shim/node/ap_footer.js
/**
 * apeman react package for footer component.
 * @class ApFooter
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApFooter */
var ApFooter = _react2.default.createClass({
  displayName: 'ApFooter',


  // --------------------
  // Specs
  // --------------------

  propTypes: {},

  mixins: [],

  statics: {},

  getInitialState: function getInitialState() {
    return {};
  },
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  render: function render() {
    var s = this;
    var props = s.props;


    return _react2.default.createElement(
      'footer',
      { className: (0, _classnames2.default)('ap-footer', props.className) },
      props.children
    );
  }
});

exports.default = ApFooter;
