// apeman-react-footer/shim/node/ap_footer_style.js
/**
 * Style for ApFooter.
 * @class ApFooterStyle
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _apemanReactStyle = require('apeman-react-style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApFooterStyle */
var ApFooterStyle = _react2.default.createClass({
  displayName: 'ApFooterStyle',

  propTypes: {
    style: _react.PropTypes.object
  },
  getDefaultProps: function getDefaultProps() {
    return {
      style: {}
    };
  },

  statics: {
    styleData: function styleData(config) {
      return {
        all: {
          '.ap-footer': {
            display: 'block',
            padding: '24px'
          }
        }
      };
    }
  },
  render: function render() {
    var s = this;
    var props = s.props;

    var _ApFooterStyle$styleD = ApFooterStyle.styleData(props);

    var all = _ApFooterStyle$styleD.all;
    var small = _ApFooterStyle$styleD.small;
    var medium = _ApFooterStyle$styleD.medium;
    var large = _ApFooterStyle$styleD.large;


    return _react2.default.createElement(
      _apemanReactStyle.ApStyle,
      { data: Object.assign(all, props.style),
        smallMediaData: small,
        mediumMediaData: medium,
        largeMediaData: large
      },
      props.children
    );
  }
});

exports.default = ApFooterStyle;
